core/src/main/java/org/elasticsearch/index/mapper/Mapper.java

```diff
@@ -177,4 +177,11 @@ public final String simpleName() {
     /** Return the merge of {@code mergeWith} into this.
      *  Both {@code this} and {@code mergeWith} will be left unmodified. */
     public abstract Mapper merge(Mapper mergeWith, boolean updateAllTypes);
+
+    /**
+     * Update the field type of this mapper. This is necessary because some mapping updates
+     * can modify mappings across several types. This method must return a copy of the mapper
+     * so that the current mapper is not modified.
+     */
+    public abstract Mapper updateFieldType(Map<String, MappedFieldType> fullNameToFieldType);
```

Comment:
There are a couple things I don't like about this:
* It is an api that is not relevant to how the mapper operates on data, but on logistics of mappings and indexes, thus complicating all mapper implementations.
* It is more complicated than what MappedFieldTypeReference was before, which was also more hidden: implementations of look complicated eg for geo mappers.

I would rather have an api for mapper immutability that was opaque to mapper implementations. However, I also don't know exactly what that would look like...but adding a new api to mappers like this doesn't seem like what we want long term?